### PR TITLE
stm: don't propagate side effects when editing a proof

### LINKED ISSUE #7088: Coq crashes when inserting `Transparent` in a proof.
To reproduce (on coq/coq@01b7de3a673eb89cea61442c4db721aad9520c9f for example), open CoqIDE on a file containing:

```
Lemma foo : True.
Proof.
(* Transparent id. *)
exact 3.
Qed.
```

then "go to end". Uncomment the `Transparent` command, change `exact 3` into `exact I` and "go to end" again.

@gares is working on a fix

### PR DESCRIPTION
**Pull request by gares**
Fix #7088 

## Problem

When a proof id edited, the only changes that are allowed are local to the proof branch being `Edit`ed.
The API `propagate_side_effect` alters all branches, not just the current one (it is its sole purpose)

## Fix

We don't propagate the side effect when we are editing a proof.

## Drawback (unfixable)

A command like `Transparent` that is added late (while editing a proof) has only effect inside the proof.
This is not the case for the same command when it is there from the very beginning.
The only fix would be to change the semantics: commands inside proofs always have a local effect, but this is for Coq 9 (for sure, not for a last minute bugfix)

### CODE REVIEW

**Comment by Zimmi48** [comment]
> The only fix would be to change the semantics: commands inside proofs always have a local effect, but this is for Coq 9

Do we have to wait for Coq 9 which we don't even know when it will arrive? Isn't the solution you propose what @ejgallego has been saying we should do for a while already? Can't it be done in 8.9? Do you think it will break too much stuff?

**Comment by ejgallego** [comment]
In principle I am Ok with the fix @gares proposes, but I want to do a deeper review. @gares, can you assign the PR to the person you'd like to merge it.

@Zimmi48 , we have some plans for 8.9 including making this behavior user controllable. There is an almost finished CEP sitting in my hard drive summarizing the discussion we had at Nice long time ago [sorry folks for the delay]

On the other hand, modifying the proof scheduler is a very hard task, also I'd like to push the `Proof_global` cleanup before it and that one is also uh, tricky. Let's see what happens but there is hope.

**Comment by maximedenes** [comment]
@ejgallego You're the secondary maintainer of the STM, so you should proceed to the merge (after you review and approve the patch).

**Comment by gares** [comment]
@Zimmi48 I'd change the behavior on the spot. But I think the change would make more sense together with a full redesign of the document. I don't think it would change drastically what is accepted, but simplify a lot the STM while providing a decent alternative to all the use cases that are today supported by the STM via hacks.

**Comment by ejgallego** [comment]
@maximedenes well @gares who is the maintainer assigned you so I guess you should take care?

**Review by ejgallego — APPROVED** [review]

**Comment by ejgallego** [comment]
> That's not how the process work, so unless there's any concrete reason, I suggest you merge instead. If any doubt, please re-read dev/doc/MERGING.md and ask for clarifications where needed.

Sure this patch is a nit, but however I do think that if the maintainer assigns a different person, he's overriding the default "pecking" order, isn't it? [Situation which I guess makes sense in some occasions]

**Comment by maximedenes** [comment]
> Sure this patch is a nit, but however I do think that if the maintainer assigns a different person, he's overriding the default "pecking" order, isn't it? [Situation which I guess makes sense in some occasions]

I don't know if it makes sense in some situations, but in the present case, I suggest you do the merge :)

**Comment by ejgallego** [comment]
> I don't know if it makes sense in some situations, but in the present case, I suggest you do the merge :)

Sure, no problem, but I'd be good to understand what the assignment meant [for future occasions]

**Comment by Zimmi48** [comment]
In principle (according to the doc), the assignment is a self-assignment anyways. But maybe indeed the doc is not clear for the case when the principal maintainer is the author of the PR.

**Comment by ejgallego** [comment]
> In principle (according to the doc), the assignment is a self-assignment anyways. But maybe indeed the doc is not clear for the case when the principal maintainer is the author of the PR.

Yup but here it was a bit of a special thing as before I could self-assign the maintainer assigned differently; I dunno XDD

**Comment by Zimmi48** [comment]
He did because you asked him to XD:

> @gares, can you assign the PR to the person you'd like to merge it.

**Comment by maximedenes** [comment]
> On the other hand, modifying the proof scheduler is a very hard task, also I'd like to push the Proof_global cleanup before it and that one is also uh, tricky. Let's see what happens but there is hope.

Are you sure that making all commands have effects only locally to a proof requires a preliminary clean up? It would be nice to soon have a patch that implements it so that we can anticipate compatibility issues.

**Comment by gares** [comment]
My understanding was that Elimio could not do the merge hence I assigned it to @maximedenes because he was pestered by the bug

**Comment by maximedenes** [comment]
I wanted to see how the test failed on `master`. I got:

```
fake_id: error: Currently, the parsing api only supports parsing at the tip of the document.
                You wanted to parse at: 1 but the current tip is: 8
```

How is that testing the bug I reported?

**Comment by gares** [comment]
I get this one, but also "Arguments... : assert" to fail. The warning says that the DAG is srewed up (the real tip has an extra node, but the "edit proof" part of the protocol does not support that scenario. But this is not an error, so it does not make fake_ide fail. What the tests does is to verify that the side effect injected into the proof at proof-fixing time is not visible outside the proof. This fails hard on master and works in the branch (at least, this is what I saw yesterday, but it was late I admit)

**Comment by gares** [comment]
Another way to put it: the test checks the new spec, rather than your exact bug.

**Comment by gares** [comment]
> Are you sure that making all commands have effects only locally to a proof requires a preliminary clean up?

Such a patch is as simple as commenting out `propagate_side_effect` and always add `VCS.commit` only in the current branch (as this PR does in the `'Edit` branch). I guess @ejgallego is afraid that would interact badly with nested proofs (whose ending Qed is a side effect and is propagated). We could try a middle ground, make `Commands` local, and still support the messy nested proofs (for a while).

**Comment by maximedenes** [comment]
Here is the full log I got:

```
==========> TESTING ide/bug7088.fake <==========
Add (("Add LoadPath \"ide\". ",-1),(1,true))
GOOD (2,(Inl ,""))
Add ((" Arguments id T x : rename. ",-2),(2,true))
GOOD (3,(Inl ,""))
Add ((" Lemma foo : True. ",-3),(3,true))
GOOD (4,(Inl ,""))
Add ((" Proof. ",-4),(4,true))
GOOD (5,(Inl ,""))
Add ((" exact 3. ",-5),(5,true))
GOOD (6,(Inl ,""))
Add ((" Qed. ",-6),(6,true))
GOOD (7,(Inl ,""))
Wait 
GOOD 
Edit_at 5
GOOD Inr (5,(7,7))

 [   initial,   2] Add LoadPath "ide". 
 [          ,   3] Arguments id T x...
 [          ,   4] Lemma foo : True. 
*[      here,   5] Proof. 
Add ((" Arguments id FOO BAR : rename. ",-7),(5,true))
GOOD (9,(Inl ,""))
Add ((" exact I. ",-8),(9,true))
GOOD (10,(Inl ,""))
Add ((" Qed. ",-9),(10,true))
GOOD (7,(Inr 8,""))

Add ((" Arguments id T x : assert. ",-10),(1,true))
FAIL 0 [Currently, the parsing api only supports parsing at the tip of the document.
You wanted to parse at: 1 but the current tip is: 8]
fake_id: error: Currently, the parsing api only supports parsing at the tip of the document.
                You wanted to parse at: 1 but the current tip is: 8
==========> FAILURE <==========
    ide/bug7088.fake...Error!
```

Where do I see the `Arguments ... : assert` error message?

**Comment by gares** [comment]
I don't see it, but if you put `:rename` in place of `:assert` you get no error.

**Comment by gares** [comment]
If you want to open another issue about fake_ide not reporting the right error...

**Comment by maximedenes** [comment]
> If you want to open another issue about fake_ide not reporting the right error...

Done, see #7095.

**Comment by ejgallego** [comment]
Some developments do actually rely on side effects being propagated, so that change is a bit tricky and requires overlays. Maybe we could add a warning?

On the other hand, side effects inside proofs are not really a problem per-se, it is just that the current design [side-effect replay being tied to the document structure itself] makes them hard to support.

In general, my point of view about all this problems is that we are really dealing with an architectural problem, no matter how much we patch, we can always find cases that won't work. So I dunno how to proceed other than doing a major-ish revision at some point.

**Comment by ejgallego** [comment]
> I guess @ejgallego is afraid that would interact badly with nested proofs (whose ending Qed is a side effect and is propagated). We could try a middle ground, make Commands local, and still support the messy nested proofs (for a while).

I think the way forward here is to decouple document construction from proof checking; as long as we keep the current model we are gonna have problems one way or another.

Once a proper scheduler is in place, it can account for "propagation" without messing with the document, even if it involves some document analysis it should be O(n) [n the size of the document] and thus bearable.

**Comment by gares** [comment]
I agree, and I'm fine with decoupling the document construction with its scheduling.

@maximedenes raises an orthogonal issue: he believes that the *intended* semantics of a `Command` inside a proof is confine its effect to the proof itself. If it is true, we can patch the current code *now* to emit warnings to help people migrate to the new semantics, or to simply detect errors using CI in a branch that makes the effects confined (by commenting out propagate side effect).

**Comment by ejgallego** [comment]
> @maximedenes raises an orthogonal issue: he believes that the intended semantics of a Command inside a proof is confine its effect to the proof itself. If it is true, we can patch the current code now to emit warnings to help people migrate to the new semantics, or to simply detect errors using CI in a branch that makes the effects confined (by commenting out propagate side effect).

I can't really make my head on this one, so far it doesn't worry me a lot as the code wouldn't gain a lot due to opacity still being broken by nested proofs / "type b side effects". I recall that @JasonGross opposed to this change tho. Anything to add @JasonGross ?

I am not sure how a warning would work tho, "type a side effects" inside proofs are needed, for example, you may want to open a notation scope, use transparent, etc...

**Comment by Zimmi48** [comment]
If we decided to make all side-effects inside a proof local, this should also apply to `Qed`, shouldn't it?
But I agree with @ejgallego concern: I don't see a proper way of displaying this warning because local side-effects are still acceptable.

**Comment by ejgallego** [comment]
> If we decided to make all side-effects inside a proof local, this should also apply to Qed, shouldn't it?

`Qed` is a special side effect. In fact there are 3 different side effects XD, kernel (!!!) side effects, what the STM calls "side-effects" [commands that alter the state] and nested lemmas, which are handled differently.

IMHO all three notions should go away, but indeed we need kernel-side support for "local" entries.

**Comment by gares** [comment]
I think the test is wrong. I thought that the message we see in fake_ide was a warning caught by mistake, but it is a proper error. Well, it is still a test... but makes #7095 invalid

**Comment by ejgallego** [comment]
Umm, well I misread the log, indeed, I have a hard time seeing why `fake_ide` is not broken?

```
Add ((" Qed. ",-9),(10,true))
GOOD (7,(Inr 8,""))
Add ((" Arguments id T x : assert. ",-10),(1,true))
```
So why does `fake_ide` adds at 1 here, that doesn't make sense, it should add at 8.

**Pull request merged**
